Modules/_json.c

```diff
@@ -1502,10 +1503,15 @@ encoder_encode_key_value(PyEncoderObject *s, _PyUnicodeWriter *writer, bool *fir
         return 0;
     }
     else {
-        PyErr_Format(PyExc_TypeError,
-                     "keys must be str, int, float, bool or None, "
-                     "not %.100s", Py_TYPE(key)->tp_name);
-        return -1;
+        newobj = PyObject_CallOneArg(s->defaultfn, key);
+        if (newobj == NULL) {
+            PyErr_Format(PyExc_TypeError,
+                         "keys must be str, int, float, bool or None, "
+                         "not %.100s", Py_TYPE(key)->tp_name);
```

Comment:
`encoder_encode_key_value` will do real type checks and raise this TypeError. Would newobj being NULL mean a wrong type was returned by `s->defaultfn`, or just that the call failed somehow?